channel/state/outcome/allocation.go

```diff
@@ -105,3 +106,51 @@ var allocationsTy = abi.ArgumentMarshaling{
 		{Name: "metadata", Type: "bytes"},
 	},
 }
+
+// DivertToGuarantee returns a new Allocations, identical to the reciever but with
+// the leftDebtee's amount reduced by leftDebit,
+// the rightDebtee's amount reduced by rightDebit,
+// and a Guarantee appended for the guaranteeDestination
+func (allocations Allocations) DivertToGuarantee(
+	leftDestination types.Destination,
+	rightDestination types.Destination,
+	leftAmount *big.Int,
+	rightAmount *big.Int,
+	guaranteeDestination types.Destination,
+) (Allocations, error) {
+
+	if leftDestination == rightDestination {
+		return Allocations{}, errors.New(`debtees must be distinct`)
+	}
+
+	newAllocations := make([]Allocation, len(allocations)+1)
+	for i, allocation := range allocations {
+		newAllocations[i] = allocation
+		switch newAllocations[i].Destination {
+		case leftDestination:
+			newAllocations[i].Amount.Sub(newAllocations[i].Amount, leftAmount)
```

Comment:
I take it the introduction of `newAllocations` is meant to prevent overwrites of the calling `allocations` value, but it seems like a mixed success. Placing some logs around `DivertToGuarantee` in the test:
```
allocation_test.go:89: a prior:  [{0x000000000000000000000000000000000000000000000000000000000000000a 243 0 []} {0x000000000000000000000000000000000000000000000000000000000000000b 309 0 []}]
allocation_test.go:93: a after:  [{0x000000000000000000000000000000000000000000000000000000000000000a 238 0 []} {0x000000000000000000000000000000000000000000000000000000000000000b 304 0 []}]
allocation_test.go:94: got:      [{0x000000000000000000000000000000000000000000000000000000000000000a 238 0 []} {0x000000000000000000000000000000000000000000000000000000000000000b 304 0 []} {0x0000000000000000000000000000000000000000000000000000000000000abc 10 1 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1
```
The existing individual `Allocation`s are modified, but the slice does not include the new guarantee.

Reply:
Good catch. I added a test to catch this (which fails), and also introduces and uses `Allocation.Clone()` to make it pass.